pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/rebalance/ReplicaGroupRebalanceSegmentStrategy.java

```diff
@@ -0,0 +1,491 @@
+
+    if (sameNumInstancesPerPartition && sameNumReplicaGroup && addedServers.size() == removedServers.size()) {
+      return ReplicaGroupRebalanceType.REPLACE;
+    } else if (sameNumInstancesPerPartition) {
+      if (oldNumReplicaGroup < targetNumReplicaGroup && !isAddedServersSizeZero && isRemovedServersSizeZero
+          && addedServers.size() % targetNumInstancesPerPartition == 0) {
+        return ReplicaGroupRebalanceType.ADD_REPLICA_GROUP;
+      } else if (oldNumReplicaGroup > targetNumReplicaGroup && isAddedServersSizeZero && !isRemovedServersSizeZero
+          && removedServers.size() % targetNumInstancesPerPartition == 0) {
+        return ReplicaGroupRebalanceType.REMOVE_REPLICA_GROUP;
+      }
+    } else if (sameNumReplicaGroup) {
+      if (oldNumInstancesPerPartition < targetNumInstancesPerPartition && isRemovedServersSizeZero
+          && !isAddedServersSizeZero && addedServers.size() % targetNumReplicaGroup == 0) {
+        return ReplicaGroupRebalanceType.ADD_SERVER;
+      } else if (oldNumInstancesPerPartition > targetNumInstancesPerPartition && !isRemovedServersSizeZero
+          && isAddedServersSizeZero && removedServers.size() % targetNumReplicaGroup == 0) {
+        return ReplicaGroupRebalanceType.REMOVE_SERVER;
+      }
+    }
+    return ReplicaGroupRebalanceType.UNSUPPORTED;
+  }
+
+  private IdealState rebalanceSegments(IdealState idealState, TableConfig tableConfig,
+      ReplicaGroupPartitionAssignment replicaGroupPartitionAssignment, boolean bootstrap) {
+    Map<String, Map<String, String>> segmentToServerMapping = idealState.getRecord().getMapFields();
+    Map<String, LinkedList<String>> serverToSegments = buildServerToSegmentMapping(segmentToServerMapping, bootstrap);
+
+    // Compute added and removed servers
+    List<String> oldServerInstances = new ArrayList<>(serverToSegments.keySet());
+    List<String> serverInstances = replicaGroupPartitionAssignment.getAllInstances();
+
+    List<String> addedServers = new ArrayList<>();
+    List<String> removedServers = new ArrayList<>();
+    for (String server : oldServerInstances) {
+      if (!serverInstances.contains(server)) {
+        removedServers.add(server);
+      }
+    }
+    for (String server : serverInstances) {
+      if (!oldServerInstances.contains(server)) {
+        addedServers.add(server);
+      }
+    }
+
+    // Add servers to the mapping
+    for (String server : addedServers) {
+      serverToSegments.put(server, new LinkedList<String>());
+    }
+
+    // Remove servers from the mapping
+    for (String server : removedServers) {
+      serverToSegments.remove(server);
+    }
+
+    // Check if rebalance can cause the downtime
+    Set<String> segmentsToCover = segmentToServerMapping.keySet();
+    Set<String> coveredSegments = new HashSet<>();
+    for (Map.Entry<String, LinkedList<String>> entry : serverToSegments.entrySet()) {
+      if (!removedServers.contains(entry.getKey())) {
```

Comment:
on line 325 all servers in removedServers were removed from serverToSegments. So removedServers will never contain any of the servers that are remaining in serverToSegments keySet right?

Reply:
good catch. I removed the check.